hadoop-hdds/docs/content/design/lifecycle-configurations.md

```diff
@@ -0,0 +1,253 @@
+---
+title: AWS S3 Lifecycle Configurations
+summary: Enables users to manage lifecycle configurations for buckets, allowing automated deletion of keys based on predefined rules.
+date: 2024-04-25
+jira: HDDS-8342
+status: draft
+---
+<!--
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+   http://www.apache.org/licenses/LICENSE-2.0
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License. See accompanying LICENSE file.
+-->
+
+# Lifecycle Management
+
+## Introduction
+I encountered the need for a retention solution within my cluster, specifically the ability to delete keys in specific paths after a certain time period.   
+This requirement closely resembled the functionality provided by AWS S3 Lifecycle configurations, particularly the Expiration part ([AWS S3 Lifecycle Configuration Examples](https://docs.aws.amazon.com/AmazonS3/latest/userguide/lifecycle-configuration-examples.html)).  
+
+## Overview
+
+### Functionality
+- User should be able to create/remove/fetch lifecycle configurations for a specific S3 bucket.
+- The lifecycle configurations will be executed periodically.
+- Depending on the rules of the lifecycle configuration there could be different actions or even multiple actions. 
+- At the moment only expiration is supported (keys get deleted).
+- The lifecycle configurations supports all buckets not only S3 buckets.
```

Comment:
S3 supports lifecycle configurations only for non-directory buckets. In that case, translating the same for Ozone, we would be supporting these only for object-store buckets and not for FSO buckets. 
Do we plan to do as above or consider FSO buckets as well? 
Also, how do we plan to handle legacy buckets? As we have the ozone.om.enable.filesystem.paths config to have flexibility on bucket behaviour.

Reply:
We can support the `key prefix` for all of types of buckets, for others `Filter element` such as `object tags` and `object size` only support for the OBS buckets.
https://docs.aws.amazon.com/AmazonS3/latest/userguide/intro-lifecycle-rules.html#intro-lifecycle-rules-filter

Reply:
For FSO buckets, due to the directory structure, whether an object can be deleted depends on its subdirectories, so Lifecycle cannot perform as expected.

For the legacy buckets, I think we can support it and there is no need to distinguish between legacy buckets and OBS buckets, because the deletion operations that Lifecycle can perform do not exceed what legacy buckets can do. (But I think support for legacy buckets is a feature that can be discussed)